service/gcs/core/gcs/gcs.go

```diff
@@ -579,40 +587,77 @@ func (c *gcsCore) WaitContainer(id string) (func() int, error) {
 	return f, nil
 }
 
-// WaitProcess returns a function that can be used to successfully wait for a
-// process exit code. On error the pid was not a valid pid.
-func (c *gcsCore) WaitProcess(pid int) (func() int, error) {
+// WaitProcess returns a channel that can be used to wait for the process exit
+// code and a channel that can be used to signal when the waiter has processed
+// the code fully or decided to stop waiting.
+//
+// The second channel must be signaled in either case to keep the wait counts in
+// sync.
+//
+// On error the pid was not a valid pid and no channels will be returned.
+func (c *gcsCore) WaitProcess(pid int) (chan int, chan bool, error) {
 	c.processCacheMutex.Lock()
 	entry, ok := c.processCache[pid]
 	if !ok {
 		c.processCacheMutex.Unlock()
-		return nil, errors.WithStack(gcserr.NewProcessDoesNotExistError(pid))
+		return nil, nil, errors.WithStack(gcserr.NewProcessDoesNotExistError(pid))
 	}
 	c.processCacheMutex.Unlock()
 
-	// Safe to do without lock because this is a write once variable at creation.
+	// If we are an init process waiter increment our count for this waiter.
 	if entry.isInitProcess {
-		// We only `Add` to the writers lock for the 2nd..Nth waiters because we
-		// added the 1st writer at creation to ensure that the container exited
-		// notification wasn't written until at least one `WaitProcess` was called.
-		entry.syncRoot.Lock()
-		if entry.writersCalled {
-			entry.writersWg.Add(1)
-		} else {
-			entry.writersCalled = true
-		}
-		entry.syncRoot.Unlock()
+		entry.writersSyncRoot.Lock()
+		entry.writersWg.Add(1)
```

Comment:
I think this is my last question, but should this `Add` be conditioned on `writersCalled` being true like it was before?

Reply:
So I change the flow in order to make safer sense on timeout cases.

Every call to `WaitProcess` where it is the *init process* will increment its wait count. When it either succeeds or times out. It will indicate that its done waiting. If it indicates that its done before an exit code was given then we know it timed out and we don't do the "extra" decrement associated with the `WaitContainer` call. If it did get the exit code before the acknowledgement then we check if its the first one and do the extra decrement but not for any additional waiters so that we keep the `WaitGroup` in sync.

Reply:
Make sense! I'll approve it :)